[Android] MediaPicker: Use New Photo Picker

phunkeler opened the pull request:
### Description of Change

Use the modern [Photo picker](https://developer.android.com/training/data-storage/shared/photopicker) via the [Activity Result API](https://developer.android.com/training/basics/intents/result)'s for selecting a photo or video. 

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes [#9615](https://github.com/dotnet/maui/issues/9615)
Fixes #29076

### Demo
**Samsung Galaxy S21 (API 34):**
| As-is | To-be |
| --- | --- |
| ![9615_api34_asis](https://github.com/user-attachments/assets/b4256f4b-88b4-448a-b884-34ca3a2074a8) | ![9615_api34_tobe](https://github.com/user-attachments/assets/5b16caf7-fe60-4b31-a672-d8a59199fa87) |

**Zebra TC75 (API 25):**
| As-is | To-be |
| --- | --- |
| ![9615_api25_asis](https://github.com/user-attachments/assets/f5bcdd99-187b-4704-91e4-8501aff53a0c) | ![9615_api25_tobe](https://github.com/user-attachments/assets/fa41d89e-c558-4f38-8a58-be59b553ff5b) |

phunkeler committed:
init

phunkeler commented on src/Essentials/src/Essentials.csproj:
```diff
@@ -42,7 +42,9 @@
     <AndroidResource Include="Resources\xml\*.xml" />
+    <PackageReference Include="Xamarin.AndroidX.Activity" Version="1.8.2.1" />
     <PackageReference Include="Xamarin.AndroidX.Browser" Version="1.5.0.3" />
+    <PackageReference Include="Xamarin.AndroidX.Collection.Ktx" Version="1.4.0.1" />
```
This avoids a `Type androidx.collection.ArraySetKt is defined multiple times` exception -- similar to #20561

phunkeler commented:
@dotnet-policy-service agree

Yu-Core commented:
According to the [photopicker document](https://developer.android.com/training/data-storage/shared/photopicker)
If the photo picker isn't available on a device, the library automatically invokes the [ACTION_OPEN_DOCUMENT](https://developer.android.com/reference/kotlin/android/content/Intent#action_open_document) intent action instead.

Intent.ActionOpenDocument behaves differently from the currently used Intent.ActionGetContent

Xiaomi MIX 2S ( API 29)
| Intent.ActionGetContent | Intent.ActionOpenDocument |
| - | - |
| ![Screenshot_2024-09-22-04-06-19-126_com google android documentsui](https://github.com/user-attachments/assets/6101110f-204a-471e-bbfd-66d0d8e76e07) | ![Screenshot_2024-09-22-04-08-47-099_com google android documentsui](https://github.com/user-attachments/assets/7f9885ee-da87-465f-ad1d-b8e7dbd06ab7) |

I think the original method should be used when the photo picker isn't available on a device

`PickVisualMedia.InvokeIsPhotoPickerAvailable(Context context)` can be used to check

NiravTyler commented:
When are we getting this feature available?

arnwest commented:
Google's announced deadline for transitioning to the new Photopicker is May 28, 2015. 
Is there any chance that this functionality will be available in .NET MAUI soon?

jfversluis commented:
Hey @phunkeler sorry to keep you waiting here, are you still interested in bringing this over the finish line? Else we will take it over. Thanks!

phunkeler committed:
refactor + prior behavior

phunkeler commented:
> Hey @phunkeler sorry to keep you waiting here, are you still interested in bringing this over the finish line? Else we will take it over. Thanks!
> 
> As per this comment: [#24027 (comment)](https://github.com/dotnet/maui/pull/24027#issuecomment-2365320028) is there maybe indeed a way to detect this and maybe keep the old behavior around? At least just for a transition period?

Hi @jfversluis, I've updated this and added the check for `IsPhotoPickerAvailable`. Please feel free to modify as you guys see fit.

Below are a few samples:

| API 34 | API 28 | API 21 |
| --- | --- | --- |
|![fix_9615-api34](https://github.com/user-attachments/assets/4f9da13b-e447-4101-897c-db470259bc32)| ![fix_9615-api28](https://github.com/user-attachments/assets/d116d7bc-e7e8-4ea2-b62b-e338fd30bdac)|![fix_9615-api21](https://github.com/user-attachments/assets/febd4689-84b5-4bc2-989a-1ed85357504d)|

jfversluis commented:
/azp run

jfversluis commented:
Awesome work, thank you so much! I'll look at this asap!

christianrr commented:
@jfversluis  This bug exists for quite a while, it would be great if you could merge this quickly into the product. Many thanks!

jfversluis commented:
@christianrr it would be very helpful if you could grab the NuGets from this PR and [test it out](https://youtu.be/B2FOBjZKm9k) and let us know if everything works as expected!

jfversluis commented:
@phunkeler are you still interested in getting this in? I'm looking to modernize the MediaPicker APIs for .NET 10. If not, I will continue with this great start from you!

jfversluis commented:
/azp run MAUI-public

phunkeler commented:
> @phunkeler are you still interested in getting this in? I'm looking to modernize the MediaPicker APIs for .NET 10. If not, I will continue with this great start from you!
> 
> If you still want to do it, could you retarted and rebase this on the net10.0 branch?

Hi @jfversluis, please proceed. I won't have much time in the coming days to dedicate to this. Thank you!

rmarinho reviewed (approved):
Seems to work fine on my device

jfversluis commented:
@phunkeler thank you so much for the initial work on this! Happy that we can now count you as a official .NET MAUI contributor! :)
